tools/src/main/java/com/nedap/archie/creation/OpenEhrRmInstanceGenerator.java

```diff
@@ -0,0 +1,375 @@
+            }
+        }
+    }
+
+    private Map<String, Object> constructArchetypeDetails(String archetypeIdValue) {
+        Map<String, Object> archetypeDetails = new LinkedHashMap<>();
+        archetypeDetails.put(typePropertyName, "ARCHETYPED");
+        Map<String, Object> archetypeId = new LinkedHashMap<>();
+        archetypeId.put(typePropertyName, "ARCHETYPE_ID");
+        archetypeId.put("value", archetypeIdValue);
+        archetypeDetails.put("archetype_id", archetypeId); //TODO: add template id?
+        archetypeDetails.put("rm_version", "1.0.4");
+        return archetypeDetails;
+    }
+
+    public void addAdditionalPropertiesAtEnd(BmmClass classDefinition, Map<String, Object> result, CObject cObject) {
+        String rmTypeName = classDefinition.getType().getBaseClass().getName();
+        switch (rmTypeName.toUpperCase()) {
+            case "DV_CODED_TEXT":
+                fixCodedText(result, cObject);
+                break;
+            case "CODE_PHRASE":
+                fixCodePhrase(result, cObject);
+                break;
+            case "HISTORY":
+                fixHistory(result);
+                break;
+            case "ELEMENT":
+                fixElement(result);
+                break;
+            case "DV_MULTIMEDIA":
+                fixDvMultimedia(result);
+                break;
+            case "PARTY_RELATED":
+                FixPartyRelated(result);
+                break;
+            case "DV_PROPORTION":
+                fixDvProportion(result);
+                break;
+            case "DV_URI":
+            case "DV_EHR_URI":
+                fixDvUri(result);
+                break;
+            case "PARTY_REF":
+                fixPartyRef(result);
+                break;
+            case "DV_INTERVAL":
+                fixDvInterval(result);
+                break;
+        }
+    }
+
+    private void fixDvInterval(Map<String, Object> result) {
+        Boolean lowerUnbounded = (Boolean) result.get("lower_unbounded");
+        Boolean lowerIncluded = (Boolean) result.get("lower_included");
+        Boolean upperUnbounded = (Boolean) result.get("upper_unbounded");
+        Boolean upperIncluded = (Boolean) result.get("upper_included");
+        if(lowerUnbounded != null && lowerIncluded != null) {
+            if(lowerUnbounded) {
+                result.put("lower_included", false);
+            }
+        }
+        if(upperUnbounded != null && upperIncluded != null) {
+            if(upperUnbounded) {
+                result.put("upper_included", false);
+            }
+        }
+        if(result.get("lower") == null) {
+            result.put("lower_unbounded", true);
+        } else {
+            result.put("lower_unbounded", false);
+        }
+        if(result.get("upper") == null) {
+            result.put("upper_unbounded", true);
+        } else {
+            result.put("upper_unbounded", false);
+        }
+    }
```

Comment:
Is the order of the checks here correct? First `x_unbounded` is used to check if `x_included` needs to be false, but `x_unbounded` might later be modified by the last two checks.

Reply:
you're right. highly unlikely to go wrong, but I will fix it.